internal/controller/snapshot/snapshot_adapter.go

```diff
@@ -433,6 +435,73 @@ func (a *Adapter) EnsureAllReleasesExist() (controller.OperationResult, error) {
 	return controller.ContinueProcessing()
 }
 
+// EnsureOverrideSnapshotValid is an operation that ensure the manually created override snapshot have valid
+// digest and git source in snapshotComponents, mark it as invalid if it is invalid
+func (a *Adapter) EnsureOverrideSnapshotValid() (controller.OperationResult, error) {
+	if !gitops.IsOverrideSnapshot(a.snapshot) {
+		a.logger.Info("The snapshot was not override snapshot, skipping")
+		return controller.ContinueProcessing()
+	}
+
+	if gitops.IsSnapshotMarkedAsInvalid(a.snapshot) {
+		a.logger.Info("The override snapshot has been marked as invalid, skipping")
+		return controller.ContinueProcessing()
+	}
+
+	var err error
+	if !controllerutil.HasControllerReference(a.snapshot) {
+		a.snapshot, err = gitops.SetOwnerReference(a.context, a.client, a.snapshot, a.application)
+		if err != nil {
+			a.logger.Error(err, fmt.Sprintf("Failed to set owner reference for snapshot %s/%s", a.snapshot.Namespace, a.snapshot.Name))
+			return controller.RequeueWithError(err)
+		}
+		a.logger.Info("Owner reference has been set to snapshot")
+	}
+
+	// validate all snapshotComponents' containerImages/source in snapshot, make all errors joined
+	var errsForSnapshot error
+	for _, snapshotComponent := range a.snapshot.Spec.Components {
+		snapshotComponent := snapshotComponent //G601
+		_, err := a.loader.GetComponent(a.context, a.client, snapshotComponent.Name, a.snapshot.Namespace)
+		if err != nil {
+			a.logger.Error(err, "Failed to get component from applicaion", "application.Name", a.application.Name, "component.Name", snapshotComponent.Name)
+			_, loaderError := h.HandleLoaderError(a.logger, err, snapshotComponent.Name, a.application.Name)
+			if loaderError != nil {
+				return controller.RequeueWithError(loaderError)
+			} else {
+				errsForSnapshot = errors.Join(errsForSnapshot, fmt.Errorf("snapshotComponent %s defined in snapshot %s doesn't exist under application %s/%s", snapshotComponent.Name, a.snapshot.Name, a.application.Namespace, a.application.Name))
+				continue
```

Comment:
Why are we skipping the rest of the for loop after we find the first place of error?

Due to this, while debugging, we would not know all the reasons why a snapshot is invalid. I think this is how it would go:
* We see that Component A is missing and this is why we have marked the override snapshot as invalid
* We would notify the user that this is the reason, and they would fix it.
* Then we will go back in this for loop, but now we are failing in the lines below this, while trying to validate the image. So we'll repeat the previous step again.

We can skip this back and forth by just capturing all the errors at once and recording & logging them. Wdyt?

Reply:
This is just `continue` the rest of loop, not `break`.

Reply:
Yes, it's `continue`, that's my point because let's say that a snapshotComponent is missing the component, valid digest, and git source too. Then based on the current logic, `errsForSnapshot` will ONLY contain the **first** error message, and **skip** through the rest of the error messages.

Reply:
I think if a component doesn't exist, then checking its source doesn't make sense...

Reply:
Hmm, I still think it makes sense to show all the errors to the users in one go.
But I need your help to figure out if I'm thinking correctly here: Let's say I'm a user who doesn't know how to create override snapshots and that's why its definition is all wrong (wrong component, invalid digest, missing git source), so I, as a user, would prefer to know everything that I've written wrong rather than just knowing that I've mentioned the wrong component name. It will save me some time. Wdyt?

Reply:
Ok, let me remove the `continue` to prompt all erros, thanks.

Reply:
Thank you!